contracts/hooks/examples/FullRange.sol

```diff
@@ -0,0 +1,326 @@
+import {CurrencyLibrary, Currency} from "@uniswap/v4-core/contracts/types/Currency.sol";
+import {TickMath} from "@uniswap/v4-core/contracts/libraries/TickMath.sol";
+import {BalanceDelta, BalanceDeltaLibrary} from "@uniswap/v4-core/contracts/types/BalanceDelta.sol";
+import {IERC20Minimal} from "@uniswap/v4-core/contracts/interfaces/external/IERC20Minimal.sol";
+import {ILockCallback} from "@uniswap/v4-core/contracts/interfaces/callback/ILockCallback.sol";
+import {PoolId, PoolIdLibrary} from "@uniswap/v4-core/contracts/types/PoolId.sol";
+import {PoolKey} from "@uniswap/v4-core/contracts/types/PoolKey.sol";
+import {FullMath} from "@uniswap/v4-core/contracts/libraries/FullMath.sol";
+import {UniswapV4ERC20} from "../../libraries/UniswapV4ERC20.sol";
+import {Position} from "@uniswap/v4-core/contracts/libraries/Position.sol";
+import {FixedPoint128} from "@uniswap/v4-core/contracts/libraries/FixedPoint128.sol";
+import {FixedPoint96} from "@uniswap/v4-core/contracts/libraries/FixedPoint96.sol";
+import {FixedPointMathLib} from "solmate/utils/FixedPointMathLib.sol";
+import {ILockCallback} from "@uniswap/v4-core/contracts/interfaces/callback/ILockCallback.sol";
+import {IERC20Metadata} from "@openzeppelin/contracts/interfaces/IERC20Metadata.sol";
+import {Strings} from "@openzeppelin/contracts/utils/Strings.sol";
+
+import "../../libraries/LiquidityAmounts.sol";
+
+contract FullRange is BaseHook, ILockCallback {
+    using CurrencyLibrary for Currency;
+    using PoolIdLibrary for PoolKey;
+    using SafeCast for uint256;
+
+    /// @notice Thrown when trying to interact with a non-initialized pool
+    error PoolNotInitialized();
+    error TickSpacingNotDefault();
+
+    /// @dev Min tick for full range with tick spacing of 60
+    int24 internal constant MIN_TICK = -887220;
+    /// @dev Max tick for full range with tick spacing of 60
+    int24 internal constant MAX_TICK = -MIN_TICK;
+
+    int256 internal constant MAX_INT = type(int256).max;
+    uint16 internal constant MINIMUM_LIQUIDITY = 10 ** 3;
+
+    struct CallbackData {
+        address sender;
+        PoolKey key;
+        IPoolManager.ModifyPositionParams params;
+    }
+
+    struct PoolInfo {
+        bool owed;
+        address liquidityToken;
+    }
+
+    mapping(PoolId => PoolInfo) public poolInfo;
+
+    constructor(IPoolManager _poolManager) BaseHook(_poolManager) {}
+
+    modifier ensure(uint256 deadline) {
+        require(deadline >= block.timestamp, "Expired");
+        _;
+    }
+
+    function getHooksCalls() public pure override returns (Hooks.Calls memory) {
+        return Hooks.Calls({
+            beforeInitialize: true,
+            afterInitialize: false,
+            beforeModifyPosition: true,
+            afterModifyPosition: false,
+            beforeSwap: true,
+            afterSwap: false,
+            beforeDonate: false,
+            afterDonate: false
+        });
+    }
+
+    function addLiquidity(
```

Comment:
this function might want to take `amountAMin`, `amountbMin` so users can be wary against slippage
https://github.com/Uniswap/v3-periphery/blob/main/contracts/NonfungiblePositionManager.sol#L151